### ISSUE TITLE
Feature: Ability to alias Account/ARN

### ISSUE THREAD

**Issue by thnee** [physera/onelogin-aws-cli#104]
This is an idea for a new feature. Do let me know if you think this is a good idea or not.

## Description

When you have multiple AWS accounts, and multiple Roles with the same name under each AWS account, the list becomes quite hard to parse for a normal human user. I suggest being able to alias the account-part of each role.

## Example

This is a realistic example at our company. In this case there are 3 AWS accounts, each one contains a role called `Admin`, plus some other roles.

```
Onelogin Password:
Google Authenticator Token: xxxxxx
Pick a role:
[1] arn:aws:iam::xxxxxxxxxxxx:role/Admin
[2] arn:aws:iam::xxxxxxxxxxxx:role/SomethingLessThanAdmin
[3] arn:aws:iam::yyyyyyyyyyyy:role/Admin
[4] arn:aws:iam::yyyyyyyyyyyy:role/SomethingLessThanAdmin
[5] arn:aws:iam::yyyyyyyyyyyy:role/SomethingLessThanAdmin2
[6] arn:aws:iam::zzzzzzzzzzzz:role/Admin
[7] arn:aws:iam::zzzzzzzzzzzz:role/SomethingLessThanAdmin
```

In order for the user to know which of the three `Admin` roles to pick, they have to reverse look up each AWS account ID in some external table. This is quite annoying and time consuming.

## Suggestion

It  would help if the program could say a human friendly string like `company-dev/Admin`, `company-stage/Admin`, and `company-prod/Admin` instead.

## Possible implementation

We could have a logic that says something like: If there is an alias for the account, display `"{account_alias}/{role_name}"`, else display `"{account_id}/{role_name}"` or just `"{role_arn}"`.

Maybe also make this feature optional with a config flag?

Maybe also provide some sorting options via config?

The only question is how to get the account alias.

There is the [ListAccountAliases](https://docs.aws.amazon.com/IAM/latest/APIReference/API_ListAccountAliases.html) API endpoint that retrieves the already configured alias for each account. However, it requires authentication.

We could just have a mapping in the config file. It could look something like this.

```
[defaults]
username = asdf

  [account-xxxxxxxxxxxx]
    alias = company-dev
  [account-yyyyyyyyyyyy]
    alias = company-stage
  [account-zzzzzzzzzzzz]
    alias = company-prod
```

This format also opens up for the possiblity to have other options per account.

**Comment by thnee** [comment]
That is even better! Thank you!

**Comment by slycoder** [comment]
Hi @e-moshaya , if not by updating the config file, what process would you prefer to retrieve account aliases?

**Comment by e-moshaya** [comment]
@slycoder please check screenshot attached. we will currently have 8 AWS accounts and this number is only going to grow over time. Currently, when I run onelogin-aws-login, I get a list of all Role ARNs as a linear list and is really difficult to ascertain which account alias the role arns fall under. The only way is to memorise the account IDs in the role arns and remember which accounts they fall under. Whereas when I login via the onelogin portal in the browser as seen in the screenshot, you can clearly see which account alias all the role arns fall under which is clear then which role relates to which account alias. It would be really nice if via the cli you also get get the alias > role ARNs hierarchy e.g.:

Account Alias 1: build
  [1] arn:aws:iam::978434673407:role/OneLogin-Read-Only-Role
  [2] arn:aws:iam::978434673407:role/OneLogin-Admin-Role
Account Alias 2: dev
  [3] arn:aws:iam::643888448669:role/OneLogin-Admin-Role
  [4] arn:aws:iam::643888448669:role/OneLogin-Read-Only-Role
Account Alias 3: staging
  [5] arn:aws:iam::947959730194:role/OneLogin-Admin-Role
  [6] arn:aws:iam::947959730194:role/OneLogin-Read-Only-Role
Account Alias 4: test
  [7] arn:aws:iam::411484814854:role/OneLogin-Admin-Role
  [8] arn:aws:iam::411484814854:role/OneLogin-Read-Only-Role
Account Alias 5: qa
  [9] arn:aws:iam::455787128138:role/OneLogin-Admin-Role
  [10] arn:aws:iam::455787128138:role/OneLogin-Read-Only-Role
Account Alias 6: master
  [11] arn:aws:iam::325708315621:role/OneLogin-Admin-Role
  [12] arn:aws:iam::325708315621:role/OneLogin-Billing-Role
  [13] arn:aws:iam::325708315621:role/OneLogin-Read-Only-Role
Account Alias 7: dev-sandbox
  [14] arn:aws:iam::125907653514:role/OneLogin-Admin-Role
Account Alias 8: prod
  [15] arn:aws:iam::125907653514:role/OneLogin-Read-Only-Role
  [16] arn:aws:iam::062554969173:role/OneLogin-Admin-Role
  [17] arn:aws:iam::062554969173:role/OneLogin-Read-Only-Role

Thoughts?

![screenshot 2019-01-04 at 15 50 39](https://user-images.githubusercontent.com/44640934/50697217-801bda00-1039-11e9-9a45-fc6cc03d80ec.png)
![screenshot 2019-01-04 at 15 44 48](https://user-images.githubusercontent.com/44640934/50697218-801bda00-1039-11e9-90ea-ef21da943f1b.png)

**Comment by slycoder** [comment]
So we get the roles out of the aws saml assertion:
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_saml_assertions.html

Do you have a sense of how to get the hierarchy you're describing?

**Comment by e-moshaya** [comment]
You can get the account alias via the account id via aws cli https://docs.aws.amazon.com/cli/latest/reference/iam/list-account-aliases.html

OneLogin does this via an external role:

with the following policy attached:

{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": [
                "iam:ListAccountAliases",
                "iam:ListRoles"
            ],
            "Effect": "Allow",
            "Resource": "*"
        }
    ]
}

https://onelogin.service-now.com/kb_view_customer.do?sys_kb_id=KB0010344#mapping

**Comment by e-moshaya** [comment]
@thnee thanks for your input! I understand the chicken and egg scenario but the workaround is to have the option to at least specify a list of multi account IAM external roles maintained in the config file. These roles would only have the policy to list roles and account aliases. onelogin-aws-cli tool could then assume these roles to list all the aliases and roles. OneLogin Config for multi account app has a configuration to allow you add an external role name and a list of salm providers ARNs. Not sure but in the meantime we're just going to have to maintain the config role_arn for aliases which is not ideal

**Comment by thnee** [comment]
I'm afraid I don't really understand what all that means. If you have a concrete suggestion for an implementation, let's see it.

I do agree that it could still technically be an improvement for some use cases to implement account aliases, at least via the config file. As I said originally, the main concern that I see is that users have to reverse look up each AWS account ID in some external table. However, since using config names/sections with `role_arn` has worked for us, I felt that it would be over engineering things, hence no PR from me so far.

**Comment by thedannywilcox** [comment]
Apologies if I should have created a new ticket instead of commenting on an old one.

I too have been looking for this feature. Doing the account through the above listed way (hard-coding arns) is not ideal due to several factors.

What about doing something like the OneLogin Assume Role library?

https://github.com/onelogin/onelogin-python-aws-assume-role/blob/master/accounts.yaml.template

Ultimately we would have a hard coded list of accounts in the accounts.yml and print the names next to it, if they are available.
